File: backend/src/main/java/dev/anthonybruno/gymbuddy/auth/AuthenticationController.java
package dev.anthonybruno.gymbuddy.auth;

import dev.anthonybruno.gymbuddy.exception.BadRequestException;
import dev.anthonybruno.gymbuddy.user.model.UserDetails;
import dev.anthonybruno.gymbuddy.util.json.Json;
import io.javalin.http.Context;

import static dev.anthonybruno.gymbuddy.auth.SessionUtils.getSession;
import static dev.anthonybruno.gymbuddy.auth.SessionUtils.setSession;

public class AuthenticationController {


    private final AuthenticationService authenticationService;

    public AuthenticationController() {
        this(new DbAuthenticationService());
    }

    public AuthenticationController(AuthenticationService authenticationService) {
        this.authenticationService = authenticationService;
    }

    public void login(Context context) {
        UserCredentials credentials = Json.intoClass(context.body(), UserCredentials.class);
        if (credentials == null) {
            throw new IllegalStateException("Need email and password");
        }
        UserDetails userDetails = this.authenticationService.login(credentials.getEmail(), credentials.getPassword())
                .orElseThrow(() -> new BadRequestException("Incorrect details"));

        setSession(context, userDetails);
        context.status(200);
        context.json(userDetails);
    }

    public void logout(Context context) {
        setSession(context, UserDetails.NOOP);
        context.status(200);
    }

    public void logCheck(Context context) {
        UserDetails sessionDetails = getSession(context);
        if (sessionDetails == null || sessionDetails == UserDetails.NOOP) {
            context.status(401);
        } else {
            context.json(getSession(context));
        }
    }


}

File: backend/src/main/kotlin/dev/anthonybruno/gymbuddy/auth/AuthenticationController.kt
package dev.anthonybruno.gymbuddy.auth

import dev.anthonybruno.gymbuddy.exception.BadRequestException
import dev.anthonybruno.gymbuddy.user.noopUserDetails
import dev.anthonybruno.gymbuddy.util.getSession
import io.javalin.http.Context


class AuthenticationController(private val authenticationService: AuthenticationService) {

    fun login(context: Context) {
        val (email, password) = context.bodyAsClass(UserCredentials::class.java)
        val userDetails = this.authenticationService.login(email, password)
                ?: throw BadRequestException("Incorrect details")

        SessionUtils.setSession(context, userDetails)
        context.status(200)
        context.json(userDetails)
    }

    fun logout(context: Context) {
        SessionUtils.setSession(context, noopUserDetails)
        context.status(200)
    }

    fun logCheck(context: Context) {
        val sessionDetails = context.getSession()
        if (sessionDetails == null || sessionDetails === noopUserDetails) {
            context.status(401)
        } else {
            context.json(sessionDetails)
        }
    }


}
